src/js/core.js

```diff
@@ -1015,35 +1013,38 @@ function MediumEditor(elements, options) {
                 this.saveSelection();
                 // Select all of the contents before calling the action
                 this.selectAllContents();
-                this.execAction(match[1], e);
+                this.execActionInternal(match[1]);
                 // Restore the previous selection
                 this.restoreSelection();
-                return;
+            } else {
+                this.execActionInternal(action);
             }
 
+            this.checkSelection();
+        },
+
+        execActionInternal: function (action) {
+            /*jslint regexp: true*/
+            var appendAction = /^append-(.+)$/gi,
+                match;
+            /*jslint regexp: false*/
+
             // Actions starting with 'append-' should attempt to format a block of text ('formatBlock') using a specific
             // type of block element (ie append-blockquote, append-h1, append-pre, etc.)
             match = appendAction.exec(action);
             if (match) {
                 this.execFormatBlock(match[1]);
-                this.setToolbarPosition();
-                this.setToolbarButtonStates();
                 return;
```

Comment:
in the scenario above, it would also make sense to `return` from the `action === 'image'` and `else` case below too.